### PR TITLE
Fix Engulf on Master

### PR DESCRIPTION
**Pull request by tserong**
Fixes: #1183, and a couple of other things. Probably shouldn't merge this until I've written some tests and added them to this PR.

Note that we also need https://github.com/SUSE/ceph/pull/207 for this to operate correctly with python3.

-----------------

**Checklist:**
- [x] Added unittests and or functional tests
- [ ] Adapted documentation
- [x] Referenced issues or internal bugtracker
- [ ] Ran integration tests successfully (trigger with "@susebot run teuthology" in a GitHub comment; see the [wiki](https://github.com/SUSE/DeepSea/wiki/Testing) for more information)

### CODE REVIEW

**Comment by tserong** [comment]
@susebot run teuthology

**Comment by tserong** [comment]
The functional test is still WIP at this point (if you're interested, check the comments in pillar-verify.sls for why it can't succeed yet)

**Comment by smithfarm** [comment]
@kshtsk Base container needs to be rebuilt?

```
Step 3/3 : RUN zypper addrepo https://download.opensuse.org/repositories/filesystems:/ceph:/mimic/openSUSE_Leap_15.0/filesystems:ceph:mimic.repo     && zypper --non-interactive --gpg-auto-import-keys refresh     && zypper --non-interactive update     && zypper --non-interactive install --no-recommends           git-core           jq           make           python3-pip           python3-tox           python3-virtualenv           rpm-build           salt-master salt-minion salt-api           sudo           tar           vim
 ---> Running in 54446f42ec19
Adding repository 'filesystems:ceph:mimic (openSUSE_Leap_15.0)' [......done]
Repository 'filesystems:ceph:mimic (openSUSE_Leap_15.0)' successfully added

URI         : http://download.opensuse.org/repositories/filesystems:/ceph:/mimic/openSUSE_Leap_15.0/
Enabled     : Yes                                                                                   
GPG Check   : Yes                                                                                   
Autorefresh : No                                                                                    
Priority    : 99 (default priority)                                                                 

Repository priorities are without effect. All enabled repositories share the same priority.
Retrieving repository 'filesystems:ceph:mimic (openSUSE_Leap_15.0)' metadata [.............

Automatically importing the following key:

  Repository:       filesystems:ceph:mimic (openSUSE_Leap_15.0)             
  Key Name:         filesystems OBS Project <filesystems@build.opensuse.org>
  Key Fingerprint:  B1FB5374 87204722 05FA6019 98C97FE7 324E6311            
  Key Created:      Tue Jul 19 14:36:42 2016                                
  Key Expires:      Thu Sep 27 14:36:42 2018 (EXPIRED)                      
  Rpm Name:         gpg-pubkey-324e6311-578e3afa                            


Warning: The gpg key signing file 'repomd.xml' has expired.
  Repository:       filesystems:ceph:mimic (openSUSE_Leap_15.0)             
  Key Name:         filesystems OBS Project <filesystems@build.opensuse.org>
  Key Fingerprint:  B1FB5374 87204722 05FA6019 98C97FE7 324E6311            
  Key Created:      Tue Jul 19 14:36:42 2016                                
  Key Expires:      Thu Sep 27 14:36:42 2018 (EXPIRED)                      
  Rpm Name:         gpg-pubkey-324e6311-578e3afa                            
........
Abort, retry, ignore? [a/r/i/...? shows all options] (a): a
Download (curl) error for 'http://download.opensuse.org/repositories/filesystems:/ceph:/mimic/openSUSE_Leap_15.0/repodata/9b118705a511c9f9111b5af10422dbbcba74fe2c18ed1c71dfea77255b8108eb-primary.xml.gz':
Error code: Curl error 47
Error message: Maximum (6) redirects followed

Repository 'filesystems:ceph:mimic (openSUSE_Leap_15.0)' is invalid.
[filesystems_ceph_mimic|http://download.opensuse.org/repositories/filesystems:/ceph:/mimic/openSUSE_Leap_15.0/] Valid metadata not found at specified URL
Please check if the URIs defined for this repository are pointing to a valid repository.
Skipping repository 'filesystems:ceph:mimic (openSUSE_Leap_15.0)' because of the above error.
error]
```

**Comment by kshtsk** [comment]
> @kshtsk Base container needs to be rebuilt?
> 
> ```
> .
> .
> .
> 'http://download.opensuse.org/repositories/filesystems:/ceph:/mimic/openSUSE_Leap_15.0/repodata/9b118705a511c9f9111b5af10422dbbcba74fe2c18ed1c71dfea77255b8108eb-primary.xml.gz':
> Error code: Curl error 47
> Error message: Maximum (6) redirects followed
> 
> Repository 'filesystems:ceph:mimic (openSUSE_Leap_15.0)' is invalid.
> [filesystems_ceph_mimic|http://download.opensuse.org/repositories/filesystems:/ceph:/mimic/openSUSE_Leap_15.0/] Valid metadata not found at specified URL
> Please check if the URIs defined for this repository are pointing to a valid repository.
> Skipping repository 'filesystems:ceph:mimic (openSUSE_Leap_15.0)' because of the above error.
> error]
> ```
It means the base is being rebuilt, but there was a connection issue to the download.opensuse.org, it means the url http://download.opensuse.org/repositories/filesystems:/ceph:/mimic/openSUSE_Leap_15.0/ is not reliable.

Just restart testing.

**Comment by kshtsk** [comment]
retest this please

**Comment by kshtsk** [comment]
@smithfarm Nbg experienced network outage I guess.

**Comment by kshtsk** [comment]
retest this please

**Comment by tserong** [comment]
@susebot run teuthology

**Comment by kshtsk** [comment]
@susebot run teuthology

**Comment by smithfarm** [comment]
@tserong Correct, not related to this PR. It's is an OVH issue we are currently investigating.

**Comment by tserong** [comment]
OK, thanks.

**Comment by kshtsk** [comment]
@susebot run teuthology

**Comment by tserong** [comment]
I'm happy with the state of this thing now if anyone would care to review? Teuthology is still failing (OVH issue still, I assume), but the functests run fine for me here.

**Comment by jschmid1** [comment]
@kshtsk @smithfarm 

This indeed looks like a teuthology issue: 

> 2018-10-06T00:00:47.742 DEBUG:teuthology.orchestra.connection:{'username': u'ubuntu', 'hostname': u'54.37.30.29', 'timeout': 240}
2018-10-06T00:00:47.852 DEBUG:teuthology.openstack:cloud_init_wait AuthenticationException Authentication failed.

File: srv/modules/runners/populate.py [hunk 1]
```diff
@@ -1101,22 +1120,22 @@ def engulf_existing_cluster(**kwargs):
...
+    has_storage_profiles = False
+
     for minion, info in local.cmd(search, "cephinspector.inspect", [],
                                   tgt_type="compound").items():
 
         if type(info) is not dict:
-            print("cephinspector.inspect failed on %s: %s" % (minion, info))
-            return False
+            return _runtime_error(exception,
+                                  "cephinspector.inspect failed on %s: %s" % (minion, info))
```
Comment (jschmid1) [inline srv/modules/runners/populate.py]: also more pythonic:

``` py
if not isinstance(info, dict):
    return _runtime_error(exception,
                          "cephinspector.inspect failed on {}: {}".format(minion, info))
```

File: srv/modules/runners/populate.py [hunk 2]
```diff
@@ -1246,26 +1261,23 @@ def engulf_existing_cluster(**kwargs):
                                 ["key=mgr." + i],
                                 tgt_type="compound")[admin_minion]
         if not mgr_keyring:
-            print("Could not obtain mgr." + i + " keyring")
-            return False
+            return _runtime_error(exception, "Could not obtain mgr." + i + " keyring")
```
Comment (jschmid1) [inline srv/modules/runners/populate.py]: also please use `.format` for string manipulation for the sake of consistency

**Comment by kshtsk** [comment]
@jschmid1 it's openstack issues, I've just switched created a CI server in another region

**Comment by kshtsk** [comment]
@susebot run teuthology

**Comment by kshtsk** [comment]
@susebot run teuthology

**Comment by jschmid1** [comment]
Still not right ->

> 12:03:37 ++ echo 'The deepsea task hasn'\''t been running. Probably teuthology issue'

**Comment by tserong** [comment]
Apparently it failed to install the ceph-test package (although I can't tell why from the logs)

**Comment by smithfarm** [comment]
It's not a problem with ceph-test. There seem to be several problems:

First, **all** repos are getting priority 1 (should be just deepsea). So zypper cries: "Repository priorities are without effect. All enabled repositories share the same priority."

Second, ceph-ci.suse.de cannot be resolved (transient DNS issue?)

```
2018-10-08T09:57:11.183 INFO:teuthology.orchestra.run.target158069067133.stdout:Retrieving repository 'deepsea' metadata [.
2018-10-08T09:57:11.184 INFO:teuthology.orchestra.run.target158069067133.stdout:Abort, retry, ignore? [a/r/i/...? shows all options] (a): a
2018-10-08T09:57:11.184 INFO:teuthology.orchestra.run.target158069067133.stdout:error]
2018-10-08T09:57:11.184 INFO:teuthology.orchestra.run.target158069067133.stdout:Warning: Skipping repository 'deepsea' because of the above error.
2018-10-08T09:57:11.184 INFO:teuthology.orchestra.run.target158069067133.stderr:Download (curl) error for 'http://ceph-ci.suse.de/artifacts/jenkins/deepsea-pr/suse:tier1/194/repodata/repomd.xml':
2018-10-08T09:57:11.184 INFO:teuthology.orchestra.run.target158069067133.stderr:Error code: Connection failed
2018-10-08T09:57:11.185 INFO:teuthology.orchestra.run.target158069067133.stderr:Error message: Could not resolve host: ceph-ci.suse.de
```

**Comment by kshtsk** [comment]
I confirm it is infrastructural problems

**Comment by kshtsk** [comment]
@susebot run teuthology

**Comment by tserong** [comment]
Excellent! That failure was because the **test** failed! :-D

````
RuntimeError: role mismatch (['admin', 'master', 'mds', 'mgr', 'mon', 'rgw', 'storage'] vs ['master', 'mds', 'mgr', 'mon', 'rgw', 'storage'])
````

Guess I'd better figure out why, and fix it.

**Comment by tserong** [comment]
So that was failing because the policy.cfg in the CI specifies both admin and master roles for the master, but the engulf will only assign the master role to the master, it won't assign the admin role (the admin role is redundant, as master already includes the admin keyring). I've updated the test to ignore this case.

**Comment by tserong** [comment]
@susebot run teuthology

**Comment by kshtsk** [comment]
@tserong Congrats!

**Comment by jschmid1** [comment]
SUSE/ceph#207 is merged. Tests have passed. Reviews addressed. :1st_place_medal:

**Pull request merged**